### PR TITLE
Split quantity class

### PR DESCRIPTION
**Pull request by clegaspi**
Splitting it up to handle different kinds of data, hopefully, more elegantly.

### CODE REVIEW

**Comment by mkhorton** [comment]
This looks really nice! Only thing I’m not sure about is the .factory method from a usability point of view, but I have an idea for a small change to move this that would make it better — can discuss tomorrow (and maybe this is fine anyway).

**Comment by clegaspi** [comment]
Something like renaming the class to "BaseQuantity" or something like that, and then creating a Quantity() function which contains the factory? That way the call looks the same as before?

**Comment by mkhorton** [comment]
Yup exactly that :)

**Comment by clegaspi** [comment]
I might get a handle on this programming thing yet! Haha

**Comment by mkhorton** [comment]
I see a lot of commits and new classes here, could you summarize what's changed (in particular the new classes/what they're for) + what still needs to be done for this to be merged?

**Comment by clegaspi** [comment]
Ok, here's a summary:

- `Quantity` has been replaced by `NumQuantity` for numerical properties and `ObjQuantity` for everything else, both of which inherit from `BaseQuantity`. Functionality specific to one type or another have been moved to the correct class.
- Intelligent quantity creation is handled by a utility class, `QuantityFactory` and the method `create_quantity()` among other class methods taken from the original `Quantity` class.
- All calls to the previous `Quantity` class have been fixed, as needed. Was overall not a large endeavor, thankfully.

In addition to the above original plan, I have also implemented some storage containers for Mongo storage.
- `StorageQuantity` holds the data of a `BaseQuantity` type object, but replaces `ProvenanceElement` object with `ProvenanceStore` objects, which contain only `ProvenanceStoreQuantity` objects.
- Upon serialization, `ProvenanceStoreQuantity` objects do not output their `value`, `units`, or `uncertainty` in effort to cut down on repeated information in the database. Upon deserialization, `ProvenanceStoreQuantity` objects will still be missing the information. However, calling `ProvenanceStoreQuantity.lookup_value(lookup)` or `StorageQuantity.to_quantity(lookup)` for the containing object with `lookup` being a function or dict which takes the `internal_id` and returns a dict containing the missing data, the `ProvenanceStoreQuantity` objects will assign these data again.

Honestly, I probably should have split this into two PRs, but I think it's kind of late for that now? Maybe not. I can do that if it seems better.

All that is left is to clean up test coverage according to Coveralls and resolve Codacy issues (which I think I just did with the last commit).

**Comment by mkhorton** [comment]
In general better as separate PRs but no need to split them now. A lot of these new classes/methods are missing docstrings, but otherwise looks reasonable. Does the web app still work with the new database format/are there any builder changes still pending?

**Comment by clegaspi** [comment]
Yes, I do need to add and update docstrings, thank you for reminding me. 

And you bring up a good point regarding the website. I will build a test database to check it out. Likely there will need to be some adjustments.

**Comment by mkhorton** [comment]
I don't think the website uses the provenance much right now, so will likely still work as long as you can deserialize a StorageQuantity into a Quantity(?) It does sound like we might want a very basic database shim though to query/retrieve quantities and materials from a built database. Thanks for all the extra detail btw, much appreciated.
